File: CLI/aira/llm.py
"""
AIRA LLM routing for optional provider-assisted scans.

The CLI remains useful in pure static mode, but this module lets advanced users
plug AIRA into local or cloud-backed model endpoints using a normalized JSON
contract.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from typing import Any, Dict, Iterable, List, Optional
from urllib import error, parse, request


AUTO_PROVIDER_ORDER = (
    "openai-compatible",
    "ollama",
    "nvidia",
    "groq",
    "gemini",
    "openrouter",
)
MAX_ATTEMPTS_PER_PROVIDER = 2
DEFAULT_TIMEOUT_SECONDS = 45
OLLAMA_DISCOVERY_TIMEOUT_SECONDS = 5


class LLMRoutingError(RuntimeError):
    """Raised when no provider can successfully complete the request."""


@dataclass
class LLMConfig:
    provider: str = "auto"
    model: Optional[str] = None
    base_url: Optional[str] = None
    timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS
    max_context_chars: int = 120_000
    system_prompt: str = ""
    user_prompt: str = ""


def _env(*names: str) -> Optional[str]:
    for name in names:
        value = os.getenv(name)
        if value:
            return value
    return None


def _provider_model(provider: str, config: Optional[LLMConfig] = None) -> Optional[str]:
    if config and config.model and config.model != "auto":
        return config.model

    if provider == "openai-compatible":
        return _env("AIRA_OPENAI_MODEL", "OPENAI_MODEL")
    if provider == "ollama":
        return _env("AIRA_OLLAMA_MODEL", "OLLAMA_MODEL")
    if provider == "nvidia":
        return _env("AIRA_NVIDIA_MODEL", "NVIDIA_MODEL") or "stepfun-ai/step-3.7-flash"
    if provider == "groq":
        return _env("AIRA_GROQ_MODEL", "GROQ_MODEL")
    if provider == "gemini":
        return _env("AIRA_GEMINI_MODEL", "GEMINI_MODEL") or "gemini-2.5-flash"
    if provider == "openrouter":
        return _env("AIRA_OPENROUTER_MODEL", "OPENROUTER_MODEL")
    return None


def _provider_base_url(provider: str, config: Optional[LLMConfig] = None) -> Optional[str]:
    if config and config.base_url:
        return config.base_url.rstrip("/")

    if provider == "openai-compatible":
        return _env("AIRA_OPENAI_BASE_URL", "OPENAI_BASE_URL")
    if provider == "ollama":
        return (_env("AIRA_OLLAMA_HOST", "OLLAMA_HOST") or "http://127.0.0.1:11434").rstrip("/")
    if provider == "nvidia":
        return "https://integrate.api.nvidia.com/v1"
    return None


def _provider_api_key(provider: str) -> Optional[str]:
    if provider == "openai-compatible":
        return _env("AIRA_OPENAI_API_KEY", "OPENAI_API_KEY")
    if provider == "nvidia":
        return _env("AIRA_NVIDIA_API_KEY", "NVIDIA_API_KEY")
    if provider == "groq":
        return _env("AIRA_GROQ_API_KEY", "GROQ_API_KEY")
    if provider == "gemini":
        return _env("AIRA_GEMINI_API_KEY", "GEMINI_API_KEY", "GOOGLE_API_KEY")
    if provider == "openrouter":
        return _env("AIRA_OPENROUTER_API_KEY", "OPENROUTER_API_KEY")
    return None


def _is_configured(provider: str, config: Optional[LLMConfig] = None) -> bool:
    if provider == "openai-compatible":
        return bool(_provider_base_url(provider, config) and _provider_model(provider, config))
    if provider == "ollama":
        return bool(_provider_model(provider, config))
    if provider == "nvidia":
        return bool(_provider_api_key(provider))
    if provider == "groq":
        return bool(_provider_api_key(provider) and _provider_model(provider, config))
    if provider == "gemini":
        return bool(_provider_api_key(provider))
    if provider == "openrouter":
        return bool(_provider_api_key(provider) and _provider_model(provider, config))
    return False


def _request_json(
    method: str,
    url: str,
    *,
    headers: Optional[Dict[str, str]] = None,
    payload: Optional[Dict[str, Any]] = None,
    timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
) -> Dict[str, Any]:
    body = None if payload is None else json.dumps(payload).encode("utf-8")
    req = request.Request(url, data=body, headers=headers or {}, method=method)
    try:
        with request.urlopen(req, timeout=timeout_seconds) as response:
            raw = response.read().decode("utf-8", errors="replace")
            return json.loads(raw or "{}")
    except error.HTTPError as exc:
        raw = exc.read().decode("utf-8", errors="replace")
        try:
            parsed = json.loads(raw or "{}")
            message = parsed.get("error", {}).get("message") or parsed.get("message") or raw
        except Exception:
            message = raw or str(exc)
        raise LLMRoutingError(f"{exc.code}: {message}") from exc
    except error.URLError as exc:
        raise LLMRoutingError(str(exc.reason)) from exc


def _fetch_ollama_models(base_url: str, timeout_seconds: int) -> List[Dict[str, Any]]:
    data = _request_json(
        "GET",
        f"{base_url.rstrip('/')}/api/tags",
        headers={"Content-Type": "application/json"},
        timeout_seconds=timeout_seconds,
    )
    models = data.get("models")
    return models if isinstance(models, list) else []


def _ollama_snapshot(config: Optional[LLMConfig] = None) -> Dict[str, Any]:
    base_url = _provider_base_url("ollama", config) or "http://127.0.0.1:11434"
    selected_model = _provider_model("ollama", config)
    snapshot = {
        "configured": bool(selected_model),
        "model": selected_model,
        "base_url": base_url,
        "reachable": False,
        "available_models": [],
        "selected_model_available": None,
    }
    try:
        models = _fetch_ollama_models(base_url, timeout_seconds=OLLAMA_DISCOVERY_TIMEOUT_SECONDS)
    except LLMRoutingError as exc:
        snapshot["message"] = f"Ollama discovery failed: {exc}"
        return snapshot

    names = [str(model.get("name") or "").strip() for model in models if str(model.get("name") or "").strip()]
    snapshot["reachable"] = True
    snapshot["available_models"] = names
    if selected_model:
        snapshot["selected_model_available"] = selected_model in names
    snapshot["configured"] = bool(selected_model)
    return snapshot


def provider_health_snapshot(config: Optional[LLMConfig] = None) -> Dict[str, Any]:
    configured = [provider for provider in AUTO_PROVIDER_ORDER if _is_configured(provider, config)]
    ollama_snapshot = _ollama_snapshot(config)
    return {
        "ok": bool(configured),
        "recommended_provider": "openai-compatible or ollama" if configured and configured[0] in {"openai-compatible", "ollama"} else "nvidia",
        "auto_provider_order": list(AUTO_PROVIDER_ORDER),
        "configured_providers": configured,
        "static_fallback": True,
        "providers": {
            provider: {
                "configured": provider in configured,
                "model": _provider_model(provider, config),
                "base_url": _provider_base_url(provider, config),
                **(
                    {
                        "reachable": ollama_snapshot["reachable"],
                        "available_models": ollama_snapshot["available_models"],
                        "selected_model_available": ollama_snapshot["selected_model_available"],
                        **({"message": ollama_snapshot["message"]} if ollama_snapshot.get("message") else {}),
                    }
                    if provider == "ollama"
                    else {}
                ),
            }
            for provider in AUTO_PROVIDER_ORDER
        },
    }

def _parse_openai_compatible_content(data: Dict[str, Any]) -> str:
    content = ((data.get("choices") or [{}])[0].get("message") or {}).get("content")
    if isinstance(content, str):
        return content.strip()
    if isinstance(content, list):
        return "".join(part.get("text", "") for part in content if isinstance(part, dict)).strip()
    return ""


def _ensure_json_text(text: str) -> str:
    cleaned = text.replace("```json", "").replace("```", "").strip()
    json.loads(cleaned)
    return cleaned


def _build_messages(system: str, prompt: str) -> List[Dict[str, str]]:
    messages = []
    if system:
        messages.append({"role": "system", "content": system})
    messages.append({"role": "user", "content": prompt})
    return messages


def _call_openai_compatible(config: LLMConfig, provider: str) -> Dict[str, Any]:
    base_url = _provider_base_url(provider, config)
    model = _provider_model(provider, config)
    if not base_url or not model:
        raise LLMRoutingError(f"{provider} is not configured.")

    headers = {"Content-Type": "application/json"}
    api_key = _provider_api_key(provider)
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"

    payload = {
        "model": model,
        "temperature": 0,
        "response_format": {"type": "json_object"},
        "messages": _build_messages(config.system_prompt, config.user_prompt),
    }
    data = _request_json(
        "POST",
        f"{base_url}/chat/completions",
        payload=payload,
        headers=headers,
        timeout_seconds=config.timeout_seconds,
    )
    return {
        "provider": provider,
        "model": model,
        "text": _ensure_json_text(_parse_openai_compatible_content(data)),
    }


def _call_ollama(config: LLMConfig) -> Dict[str, Any]:
    base_url = _provider_base_url("ollama", config)
    model = _provider_model("ollama", config)
    if not base_url or not model:
        raise LLMRoutingError("ollama is not configured.")

    available_models = []
    try:
        available_models = [
            str(entry.get("name") or "").strip()
            for entry in _fetch_ollama_models(base_url, timeout_seconds=min(config.timeout_seconds, OLLAMA_DISCOVERY_TIMEOUT_SECONDS))
            if str(entry.get("name") or "").strip()
        ]
    except LLMRoutingError:
        available_models = []

    if available_models and model not in available_models:
        sample = ", ".join(available_models[:10])
        more = " ..." if len(available_models) > 10 else ""
        raise LLMRoutingError(f"Selected Ollama model '{model}' is not available. Available models: {sample}{more}")

    payload = {
        "model": model,
        "stream": False,
        "format": "json",
        "messages": _build_messages(config.system_prompt, config.user_prompt),
        "options": {"temperature": 0},
    }
    data = _request_json(
        "POST",
        f"{base_url}/api/chat",
        payload=payload,
        headers={"Content-Type": "application/json"},
        timeout_seconds=config.timeout_seconds,
    )
    content = ((data.get("message") or {}).get("content") or "").strip()
    return {
        "provider": "ollama",
        "model": model,
        "text": _ensure_json_text(content),
    }


def _call_nvidia(config: LLMConfig) -> Dict[str, Any]:
    base_url = "https://integrate.api.nvidia.com/v1"
    model = _provider_model("nvidia", config)
    api_key = _provider_api_key("nvidia")
    if not api_key or not model:
        raise LLMRoutingError("nvidia is not configured.")

    payload = {
        "model": model,
        "temperature": 0,
        "response_format": {"type": "json_object"},
        "messages": _build_messages(config.system_prompt, config.user_prompt),
    }
    data = _request_json(
        "POST",
        f"{base_url}/chat/completions",
        payload=payload,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
            "Accept": "application/json",
        },
        timeout_seconds=config.timeout_seconds,
    )
    return {
        "provider": "nvidia",
        "model": model,
        "text": _ensure_json_text(_parse_openai_compatible_content(data)),
    }


def _call_groq(config: LLMConfig) -> Dict[str, Any]:
    base_url = "https://api.groq.com/openai/v1"
    model = _provider_model("groq", config)
    api_key = _provider_api_key("groq")
    if not api_key or not model:
        raise LLMRoutingError("groq is not configured.")

    payload = {
        "model": model,
        "temperature": 0,
        "response_format": {"type": "json_object"},
        "messages": _build_messages(config.system_prompt, config.user_prompt),
    }
    data = _request_json(
        "POST",
        f"{base_url}/chat/completions",
        payload=payload,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        timeout_seconds=config.timeout_seconds,
    )
    return {
        "provider": "groq",
        "model": model,
        "text": _ensure_json_text(_parse_openai_compatible_content(data)),
    }


def _call_openrouter(config: LLMConfig) -> Dict[str, Any]:
    model = _provider_model("openrouter", config)
    api_key = _provider_api_key("openrouter")
    if not api_key or not model:
        raise LLMRoutingError("openrouter is not configured.")

    payload = {
        "model": model,
        "temperature": 0,
        "response_format": {"type": "json_object"},
        "messages": _build_messages(config.system_prompt, config.user_prompt),
    }
    data = _request_json(
        "POST",
        "https://openrouter.ai/api/v1/chat/completions",
        payload=payload,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        timeout_seconds=config.timeout_seconds,
    )
    return {
        "provider": "openrouter",
        "model": model,
        "text": _ensure_json_text(_parse_openai_compatible_content(data)),
    }


def _call_gemini(config: LLMConfig) -> Dict[str, Any]:
    model = _provider_model("gemini", config)
    api_key = _provider_api_key("gemini")
    if not api_key or not model:
        raise LLMRoutingError("gemini is not configured.")

    payload = {
        "systemInstruction": {"parts": [{"text": config.system_prompt}]} if config.system_prompt else None,
        "contents": [{"role": "user", "parts": [{"text": config.user_prompt}]}],
        "generationConfig": {
            "temperature": 0,
            "responseMimeType": "application/json",
        },
    }
    data = _request_json(
        "POST",
        f"https://generativelanguage.googleapis.com/v1beta/models/{parse.quote(model)}:generateContent?key={parse.quote(api_key)}",
        payload=payload,
        headers={"Content-Type": "application/json"},
        timeout_seconds=config.timeout_seconds,
    )
    text = "".join(
        part.get("text", "")
        for candidate in (data.get("candidates") or [])
        for part in ((candidate.get("content") or {}).get("parts") or [])
        if isinstance(part, dict)
    ).strip()
    return {
        "provider": "gemini",
        "model": model,
        "text": _ensure_json_text(text),
    }


def _resolved_provider_order(config: LLMConfig) -> List[str]:
    if config.provider != "auto":
        return [config.provider]
    return [provider for provider in AUTO_PROVIDER_ORDER if _is_configured(provider, config)]


def _runner_for(provider: str):
    if provider == "openai-compatible":
        return lambda cfg: _call_openai_compatible(cfg, "openai-compatible")
    if provider == "ollama":
        return _call_ollama
    if provider == "nvidia":
        return _call_nvidia
    if provider == "groq":
        return _call_groq
    if provider == "gemini":
        return _call_gemini
    if provider == "openrouter":
        return _call_openrouter
    raise LLMRoutingError(f"Unsupported provider: {provider}")


def run_llm_json_audit(config: LLMConfig, system_prompt: str, user_prompt: str) -> Dict[str, Any]:
    if not user_prompt.strip():
        raise LLMRoutingError("No scan prompt provided.")

    provider_order = _resolved_provider_order(config)
    if not provider_order:
        raise LLMRoutingError(
            "No LLM providers are configured. Set local OpenAI-compatible or Ollama settings, or configure NVIDIA/Groq/Gemini/OpenRouter."
        )

    config.system_prompt = system_prompt
    config.user_prompt = user_prompt

    errors: List[str] = []
    for provider in provider_order:
        runner = _runner_for(provider)
        for attempt in range(1, MAX_ATTEMPTS_PER_PROVIDER + 1):
            try:
                return runner(config)
            except Exception as exc:
                errors.append(f"{provider} attempt {attempt}: {exc}")

    raise LLMRoutingError("All configured providers failed. " + " | ".join(errors))

File: lib/llm.js
const AUTO_PROVIDER_ORDER = ['ollama', 'nvidia', 'groq', 'gemini', 'openrouter'];
const MAX_ATTEMPTS_PER_PROVIDER = 2;
const OLLAMA_DISCOVERY_TIMEOUT_MS = 5000;
const NVIDIA_BASE_URL = 'https://integrate.api.nvidia.com/v1';

function configuredModel(provider, overrides = {}) {
  if (typeof overrides.model === 'string' && overrides.model.trim() && overrides.model !== 'auto') {
    return overrides.model.trim();
  }
  if (provider === 'ollama') return process.env.AIRA_OLLAMA_MODEL || process.env.OLLAMA_MODEL || null;
  if (provider === 'nvidia') return process.env.AIRA_NVIDIA_MODEL || process.env.NVIDIA_MODEL || 'stepfun-ai/step-3.7-flash';
  if (provider === 'groq') return process.env.AIRA_GROQ_MODEL || process.env.GROQ_MODEL || 'llama-3.1-8b-instant';
  if (provider === 'gemini') return process.env.AIRA_GEMINI_MODEL || process.env.GEMINI_MODEL || 'gemini-2.5-flash';
  if (provider === 'openrouter') return process.env.AIRA_OPENROUTER_MODEL || process.env.OPENROUTER_MODEL || null;
  return null;
}

function providerBaseUrl(provider, overrides = {}) {
  if (typeof overrides.baseUrl === 'string' && overrides.baseUrl.trim()) {
    return overrides.baseUrl.trim().replace(/\/$/, '');
  }
  if (provider === 'ollama') {
    return (process.env.AIRA_OLLAMA_HOST || process.env.OLLAMA_HOST || 'http://127.0.0.1:11434').replace(/\/$/, '');
  }
  if (provider === 'nvidia') {
    return NVIDIA_BASE_URL;
  }
  return null;
}

function providerApiKey(provider) {
  if (provider === 'nvidia') return process.env.AIRA_NVIDIA_API_KEY || process.env.NVIDIA_API_KEY || null;
  if (provider === 'groq') return process.env.AIRA_GROQ_API_KEY || process.env.GROQ_API_KEY || null;
  if (provider === 'openrouter') return process.env.AIRA_OPENROUTER_API_KEY || process.env.OPENROUTER_API_KEY || null;
  if (provider === 'gemini') return process.env.AIRA_GEMINI_API_KEY || process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY || null;
  return null;
}

async function parseJsonResponse(response, providerName) {
  const data = await response.json().catch(() => ({}));
  if (!response.ok) {
    const message =
      data?.error?.message ||
      data?.error?.metadata?.raw ||
      data?.message ||
      `${providerName} API error ${response.status}`;
    throw new Error(message);
  }
  return data;
}

async function fetchOllamaModels(baseUrl) {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), OLLAMA_DISCOVERY_TIMEOUT_MS);
  try {
    const response = await fetch(`${baseUrl}/api/tags`, {
      method: 'GET',
      headers: { 'Content-Type': 'application/json' },
      signal: controller.signal
    });
    const data = await parseJsonResponse(response, 'Ollama');
    return Array.isArray(data?.models)
      ? data.models
          .map(model => typeof model?.name === 'string' ? model.name.trim() : '')
          .filter(Boolean)
      : [];
  } catch (error) {
    if (error?.name === 'AbortError') {
      throw new Error('Ollama discovery timed out');
    }
    throw error;
  } finally {
    clearTimeout(timeout);
  }
}

async function ollamaSnapshot(overrides = {}) {
  const baseUrl = providerBaseUrl('ollama', overrides);
  const model = configuredModel('ollama', overrides);
  const snapshot = {
    configured: Boolean(model),
    model,
    base_url: baseUrl,
    reachable: false,
    available_models: [],
    selected_model_available: null
  };

  try {
    const models = await fetchOllamaModels(baseUrl);
    snapshot.reachable = true;
    snapshot.available_models = models;
    if (model) {
      snapshot.selected_model_available = models.includes(model);
    }
  } catch (error) {
    snapshot.message = `Ollama discovery failed: ${error.message}`;
  }

  return snapshot;
}

async function configuredProviders(overrides = {}) {
  const providers = [];
  const ollama = await ollamaSnapshot(overrides);
  if (ollama.configured) providers.push('ollama');
  if (providerApiKey('nvidia')) providers.push('nvidia');
  if (providerApiKey('groq')) providers.push('groq');
  if (providerApiKey('gemini')) providers.push('gemini');
  if (providerApiKey('openrouter') && configuredModel('openrouter', overrides)) providers.push('openrouter');
  return { providers, ollama };
}

function buildMessages(system, prompt) {
  return [
    system ? { role: 'system', content: system } : null,
    { role: 'user', content: prompt }
  ].filter(Boolean);
}

function parseOpenAiCompatibleText(data) {
  const content = data?.choices?.[0]?.message?.content;
  if (typeof content === 'string') return content.trim();
  if (Array.isArray(content)) {
    return content
      .map(part => typeof part?.text === 'string' ? part.text : '')
      .join('')
      .trim();
  }
  return '';
}

function ensureValidJson(text) {
  const cleaned = text.replace(/```json|```/g, '').trim();
  JSON.parse(cleaned);
  return cleaned;
}

async function callNvidia({ system, prompt }) {
  const apiKey = providerApiKey('nvidia');
  if (!apiKey) {
    throw new Error('NVIDIA_API_KEY is not configured.');
  }

  const model = configuredModel('nvidia');
  const response = await fetch(`${NVIDIA_BASE_URL}/chat/completions`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
      'Accept': 'application/json'
    },
    body: JSON.stringify({
      model,
      temperature: 0,
      response_format: { type: 'json_object' },
      messages: buildMessages(system, prompt)
    })
  });

  const data = await parseJsonResponse(response, 'NVIDIA');
  return {
    provider: 'nvidia',
    model,
    text: ensureValidJson(parseOpenAiCompatibleText(data))
  };
}

async function callGroq({ system, prompt }) {
  const apiKey = providerApiKey('groq');
  if (!apiKey) {
    throw new Error('GROQ_API_KEY is not configured.');
  }

  const model = configuredModel('groq');
  const response = await fetch('https://api.groq.com/openai/v1/chat/completions', {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${apiKey}`,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      model,
      temperature: 0,
      response_format: { type: 'json_object' },
      messages: buildMessages(system, prompt)
    })
  });

  const data = await parseJsonResponse(response, 'Groq');
  return {
    provider: 'groq',
    model,
    text: ensureValidJson(parseOpenAiCompatibleText(data))
  };
}

async function callOpenRouter({ system, prompt }) {
  const apiKey = providerApiKey('openrouter');
  const model = configuredModel('openrouter');
  if (!apiKey || !model) {
    throw new Error('OPENROUTER_API_KEY and OPENROUTER_MODEL must both be configured.');
  }

  const response = await fetch('https://openrouter.ai/api/v1/chat/completions', {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${apiKey}`,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      model,
      temperature: 0,
      response_format: { type: 'json_object' },
      messages: buildMessages(system, prompt)
    })
  });

  const data = await parseJsonResponse(response, 'OpenRouter');
  return {
    provider: 'openrouter',
    model,
    text: ensureValidJson(parseOpenAiCompatibleText(data))
  };
}

async function callGemini({ system, prompt }) {
  const apiKey = providerApiKey('gemini');
  if (!apiKey) {
    throw new Error('GEMINI_API_KEY or GOOGLE_API_KEY is not configured.');
  }

  const model = configuredModel('gemini');
  const response = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/${encodeURIComponent(model)}:generateContent?key=${encodeURIComponent(apiKey)}`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        systemInstruction: system ? { parts: [{ text: system }] } : undefined,
        contents: [
          {
            role: 'user',
            parts: [{ text: prompt }]
          }
        ],
        generationConfig: {
          temperature: 0,
          responseMimeType: 'application/json'
        }
      })
    }
  );

  const data = await parseJsonResponse(response, 'Gemini');
  const text = (data?.candidates || [])
    .flatMap(candidate => candidate?.content?.parts || [])
    .map(part => part?.text || '')
    .join('')
    .trim();

  if (!text) {
    throw new Error('Gemini returned an empty response.');
  }

  return {
    provider: 'gemini',
    model,
    text: ensureValidJson(text)
  };
}

async function callOllama({ system, prompt, model, baseUrl }) {
  const resolvedBaseUrl = providerBaseUrl('ollama', { baseUrl });
  const resolvedModel = configuredModel('ollama', { model });
  if (!resolvedBaseUrl || !resolvedModel) {
    throw new Error('Ollama is not configured. Set AIRA_OLLAMA_MODEL or pass a model explicitly.');
  }

  const availableModels = await fetchOllamaModels(resolvedBaseUrl).catch(() => []);
  if (availableModels.length > 0 && !availableModels.includes(resolvedModel)) {
    const sample = availableModels.slice(0, 10).join(', ');
    const more = availableModels.length > 10 ? ' ...' : '';
    throw new Error(`Selected Ollama model '${resolvedModel}' is not available. Available models: ${sample}${more}`);
  }

  const response = await fetch(`${resolvedBaseUrl}/api/chat`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      model: resolvedModel,
      stream: false,
      format: 'json',
      messages: buildMessages(system, prompt),
      options: { temperature: 0 }
    })
  });

  const data = await parseJsonResponse(response, 'Ollama');
  const text = (data?.message?.content || '').trim();
  if (!text) {
    throw new Error('Ollama returned an empty response.');
  }

  return {
    provider: 'ollama',
    model: resolvedModel,
    text: ensureValidJson(text)
  };
}

const PROVIDERS = {
  ollama: callOllama,
  nvidia: callNvidia,
  groq: callGroq,
  openrouter: callOpenRouter,
  gemini: callGemini
};

export async function providerHealthSnapshot(overrides = {}) {
  const { providers, ollama } = await configuredProviders(overrides);
  return {
    ok: providers.length > 0,
    recommended_provider: providers[0] === 'ollama' ? 'ollama' : providers[0] === 'nvidia' ? 'nvidia' : 'groq',
    auto_provider_order: AUTO_PROVIDER_ORDER,
    configured_providers: providers,
    heuristic_fallback: true,
    providers: {
      ollama,
      nvidia: {
        configured: providers.includes('nvidia'),
        model: configuredModel('nvidia'),
        base_url: NVIDIA_BASE_URL
      },
      groq: {
        configured: providers.includes('groq'),
        model: configuredModel('groq')
      },
      gemini: {
        configured: providers.includes('gemini'),
        model: configuredModel('gemini')
      },
      openrouter: {
        configured: providers.includes('openrouter'),
        model: configuredModel('openrouter')
      }
    }
  };
}

export async function runLLM({ system = '', prompt = '', provider = 'auto', model = '', baseUrl = '' }) {
  if (!prompt.trim()) {
    throw new Error('No scan prompt provided.');
  }

  const { providers } = await configuredProviders({ model, baseUrl });
  const providerOrder = provider === 'auto'
    ? providers
    : [provider];

  if (providerOrder.length === 0) {
    throw new Error(
      'No providers are configured. Set AIRA_OLLAMA_MODEL or OLLAMA_MODEL for Ollama, or configure NVIDIA/Groq/Gemini/OpenRouter env vars.'
    );
  }

  const errors = [];
  for (const providerName of providerOrder) {
    const runner = PROVIDERS[providerName];
    if (!runner) {
      errors.push(`${providerName}: unsupported provider`);
      continue;
    }

    for (let attempt = 1; attempt <= MAX_ATTEMPTS_PER_PROVIDER; attempt += 1) {
      try {
        return await runner({ system, prompt, model, baseUrl });
      } catch (err) {
        errors.push(`${providerName} attempt ${attempt}: ${err.message}`);
      }
    }
  }

  throw new Error(`All configured providers failed. ${errors.join(' | ')}`);
}
